torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -1422,12 +1417,9 @@ def _register_post_backward_hook(
         "The `grad_fn` is needed to access the `AccumulateGrad` and "
         "register the post-backward hook",
     )
-    acc_grad = temp_flat_param.grad_fn.next_functions[0][0]  # type: ignore[union-attr]
-    assert acc_grad is not None
-    hook_handle = acc_grad.register_hook(
-        functools.partial(_post_backward_hook, state, handle)
-    )
-    flat_param._post_backward_hook_state = (acc_grad, hook_handle)  # type: ignore[attr-defined]
+    hook = functools.partial(_post_backward_hook, state, handle)
+    hook_handle = temp_flat_param.register_post_accumulate_grad_hook(hook)
```

Comment:
```suggestion
    hook_handle = flat_param.register_post_accumulate_grad_hook(hook)
```

And you can remove the code creating temp_flat_param above.
This is also the cause of CI failures saying you can only register this hook on leafs.